packages/webdriverio/src/commands/browser/switchFrame.ts

```diff
@@ -0,0 +1,217 @@
+ *     ```ts
+ *     // switch to first frame that contains an element with id "#frameContent"
+ *     await browser.switchFrame(() => $('#frameContent').isExisting())
+ *     ```
+ *
+ *   - If given `null` it will switch to the top level frame
+ *     ```ts
+ *     // first switch into a frame
+ *     await browser.switchFrame($('iframe'))
+ *     // do more automation within that frame, then ...
+ *
+ *     // switch to the top level frame
+ *     await browser.switchFrame(null)
+ *     ```
+ *
+ * @alias browser.switchFrame
+ * @param {string|object|function} context
+ * @returns {Promise<string>} the current active context id
+ */
+export async function switchFrame (
+    this: WebdriverIO.Browser,
+    context: WebdriverIO.Element | ChainablePromiseElement | string | null | ((tree: FlatContextTree) => boolean | Promise<boolean>)
+) {
+    /**
+     * if context is `null` the user is switching to the top level frame
+     * which is always represented by the value of `getWindowHandle`
+     */
+    if (context === null) {
+        const handle = await this.getWindowHandle()
+        await switchToFrameHelper(this, handle)
+        return handle
+    }
+
+    if (typeof context === 'string') {
+        const tree = await this.browsingContextGetTree({})
+        const urlContext = findContext(context, tree.contexts, byUrl)
+        if (urlContext) {
+            await switchToFrameHelper(this, urlContext)
+            return urlContext
+        }
+
+        const urlContextContaining = findContext(context, tree.contexts, byUrlContaining)
+        if (urlContextContaining) {
+            await switchToFrameHelper(this, urlContextContaining)
+            return urlContextContaining
+        }
+
+        const contextIdContext = findContext(context, tree.contexts, byContextId)
+        if (contextIdContext) {
+            await switchToFrameHelper(this, contextIdContext)
+            return contextIdContext
+        }
+
+        throw new Error(`Frame with url or context id "${context}" not found`)
+    }
+
+    /**
+     * we are dealing with a WebdriverIO element
+     */
+    if (typeof context === 'object' && typeof (context as WebdriverIO.Element).getElement === 'function') {
+        const element = await context.getElement()
+        return switchToFrameUsingElement(this, element)
+    }
+
+    if (typeof context === 'function') {
+        const tree = await this.browsingContextGetTree({})
+        const mapContext = (context: local.BrowsingContextInfo): any => [
+            context.context,
+            ...(context.children || []).map(mapContext)
+        ]
+        const sessionContext = getContextManager(this)
+        const currentContext = await sessionContext.getCurrentContext()
+
+        /**
+         * transform context tree into a flat list of context objects with references
+         * to children
+         */
+        const allContexts: Record<string, FlatContextTree> = tree.contexts.map(mapContext).flat(Infinity)
```

Comment:
what does the .flat(Infinite) do here? haven't seen this before

Reply:
`.flat()` by default only flattens with a depth of `1`, so:

```ts
[1, [2, [3]]].flat() === [1, 2, [3]]
```

With infinity we flatten the array infinity times:

```ts
[1, [2, [3]]].flat() === [1, 2, 3]
```